tests/unit/s2n_ktls_test.c

```diff
@@ -36,7 +47,128 @@ int main(int argc, char **argv)
+                s2n_connection_ptr_free);
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+        uint8_t key_size = server_conn->secure->cipher_suite->record_alg->cipher->key_material_size;
+        EXPECT_EQUAL(key_size, S2N_TLS_AES_128_GCM_KEY_LEN);
+
+        EXPECT_OK(s2n_ktls_validate(server_conn));
+    }
+
+    /* s2n_ktls_validate TLS 1.3 */
+    {
+        if (s2n_is_tls13_fully_supported()) {
+            DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL, s2n_cert_chain_and_key_ptr_free);
+            EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+                    S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+            EXPECT_SUCCESS(s2n_config_disable_x509_verification(config));
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+            DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+
+            EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS13);
+            uint8_t key_size = server_conn->secure->cipher_suite->record_alg->cipher->key_material_size;
+            EXPECT_EQUAL(key_size, S2N_TLS_AES_128_GCM_KEY_LEN);
+
+            EXPECT_ERROR(s2n_ktls_validate(server_conn));
```

Comment:
We've learned from painful experience that EXPECT_ERROR is not your friend. You want EXPECT_ERROR_WITH_ERRNO, and you want to check for an error other than S2N_ERR_SAFETY. Otherwise, this could just as easily be failing on a random safety check than the line you're looking for.

If you can't get specific enough with your failure reason, then you need a "control" call to your method to verify that it succeeds with only one difference. That can make you confident that you're testing what you think you're testing. Like:
```
EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS13);
EXPECT_ERROR(s2n_ktls_validate(server_conn));
server_conn->actual_protocl_version = S2N_TLS12;
EXPECT_SUCCESS(s2n_ktls_validate(server_conn));
```